pkg/receive/handler.go

```diff
@@ -482,6 +528,63 @@ func (h *Handler) receiveHTTP(w http.ResponseWriter, r *http.Request) {
 	h.writeSamplesTotal.WithLabelValues(strconv.Itoa(responseStatusCode), tenant).Observe(float64(totalSamples))
 }
 
+// isUnderLimit ensures that the current number of active series for a tenant does not exceed given limit.
+// It does so in a best-effort way, i.e, in case meta-monitoring is unreachable, it does not impose limits.
+// TODO(saswatamcode): Add capability to configure diff limits for diff tenants.
+func (h *Handler) isUnderLimit(ctx context.Context, tenant string, logger log.Logger) (bool, error) {
+	if h.options.MaxPerTenantLimit == 0 || h.options.MetaMonitoringUrl.Host == "" {
+		return true, nil
+	}
+
+	// Use specified HTTPConfig to make requests to meta-monitoring.
+	httpConfContentYaml, err := h.options.MetaMonitoringHttpClient.Content()
+	if err != nil {
+		return true, errors.Wrap(err, "getting http client config")
+	}
+
+	httpClientConfig, err := httpconfig.NewClientConfigFromYAML(httpConfContentYaml)
+	if err != nil {
+		return true, errors.Wrap(err, "parsing http config YAML")
+	}
+
+	httpClient, err := httpconfig.NewHTTPClient(*httpClientConfig, "thanos-receive")
```

Comment:
I'm not sure if we should be creating a new client per remote-write request. This can be quite wasteful in terms of memory and connection use. Can we create the client when the receiver starts?

Reply:
Absolutely. Querying periodically and caching the result is the way to go 👍

Reply:
Still need to do this!  🙂

Reply:
So in https://github.com/thanos-io/thanos/pull/5520/commits/96a821eccd0f5a25b549a492237910d37edbc988, I added the query logic inside of a `runutil.Repeat` which executes the query every 15s, and caches the result in a map in tenant_id:currentSeries format. This map is used in `isUnderLimit` method.

I was a bit worried, that in a hashring setup this might have sync issues, with a late-started Receive having outdated cache. So I added a test for this too. 🙂 

Let me know if this makes sense, or if the previous approach is preferred here!